### PR TITLE
Added capablity to customize branch protection settings

### PR DESCRIPTION
**Pull request by avraf**
This code should solve two problems with the `Repository.protect_branch` method.

1. The lack of being able to control the "Require branches to be up to date before merging" field
2. Change in the github branch protection API

`Repository.protect_branch` maps to 
 :calls: `PATCH /repos/:owner/:repo/branches/:branch <https://developer.github.com/v3/repos/#enabling-and-disabling-branch-protection>`_

If you got to the link (that was provided in docstring of `Repository.protect_branch`) above, you will notice that the section "enabling-and-disabling-branch-protection" was changed.

[Here is the new API to update branch protection](https://developer.github.com/v3/repos/branches/#update-branch-protection).

I also added a way to remove branch protection since that changed too.
[See it here](https://developer.github.com/v3/repos/branches/#remove-branch-protection)

The PATCH method you used still works (but doesn't solve problem 1 above), however, it isn't listed in the current API.

I did not change the `Repository.protect_branch` method, but added the two new methods.
This way I didn't break anything and the names of the two methods reflect the new Github API.

If I need to add more testing or change them in any way let me know.

I hope this helps.

### CODE REVIEW

**Comment by gforcada** [comment]
Any chances of getting this merged and released? I'm using pygithub 1.34 and I'm unable to protect branches right now.

**Comment by gforcada** [comment]
Sorry, I'm still able to protect branches with version 1.34, the token I used did not have enough permissions.

Still, getting this merged will make it use the newer API, so 👍 on it!

File: github/Repository.py [hunk 1]
```diff
@@ -2397,6 +2397,54 @@ def protect_branch(self, branch, enabled, enforcement_level=github.GithubObject.
             headers={'Accept': 'application/vnd.github.loki-preview+json'}
         )
 
+    def remove_branch_protection(self, branch):
+        """
+        :calls: `DELETE /repos/:owner/:repo/branches/:branch/protection <https://developer.github.com/v3/repos/branches/#remove-branch-protection>`_
+        """
+        return self._requester.requestJsonAndCheck(
+            'DELETE',
+            "{base_url}/branches/{branch}/protection".format(base_url=self.url,
```
Comment (sfdye) [inline github/Repository.py]: I personally like the `format` and even f-string, but just for the sake of being consistent, would you mind changing this to the simple "+"?

File: github/Repository.py [hunk 2]
```diff
@@ -2397,6 +2397,54 @@ def protect_branch(self, branch, enabled, enforcement_level=github.GithubObject.
...
+        """
+        if require_review_all or require_review:
+            require_review = {'include_admins': require_review_all}
+
+        required_status_checks = self._build_required_status_checks(
+                                                                required_status_checks)
+
+        return self._requester.requestJsonAndCheck(
+            'PUT',
+            "{base_url}/branches/{branch}/protection".format(base_url=self.url,
```
Comment (sfdye) [inline github/Repository.py]: And here

File: github/Repository.py [hunk 3]
```diff
@@ -2397,6 +2397,54 @@ def protect_branch(self, branch, enabled, enforcement_level=github.GithubObject.
...
 
+    def remove_branch_protection(self, branch):
+        """
+        :calls: `DELETE /repos/:owner/:repo/branches/:branch/protection <https://developer.github.com/v3/repos/branches/#remove-branch-protection>`_
+        """
+        return self._requester.requestJsonAndCheck(
+            'DELETE',
+            "{base_url}/branches/{branch}/protection".format(base_url=self.url,
+                                                                branch=branch),
+            headers={'Accept': 'application/vnd.github.loki-preview+json'}
```
Comment (sfdye) [inline github/Repository.py]: Also it seems that this header is no long required

File: github/Repository.py [hunk 4]
```diff
@@ -2397,6 +2397,54 @@ def protect_branch(self, branch, enabled, enforcement_level=github.GithubObject.
...
+        return self._requester.requestJsonAndCheck(
+            'PUT',
+            "{base_url}/branches/{branch}/protection".format(base_url=self.url,
+                                                                branch=branch),
+            input={
+                'required_status_checks': required_status_checks,
+                'required_pull_request_reviews': require_review,
+                'restrictions' : restrictions
+            },
+            headers={'Accept': 'application/vnd.github.loki-preview+json'}
```
Comment (sfdye) [inline github/Repository.py]: Here too

**Comment by arenstar** [comment]
Please merge 👍

**Pull request closed without merging**